### ISSUE TITLE
VS2017express C#, tabs replaced with spaces

### ISSUE THREAD

**Issue by vsfeedback** [dotnet/roslyn#24031]
Hello,

under the following circumstances, VS2017express replaces tabs with spaces even though I deactivated that feature in the options:
- I am working on a C# file
- I am pasting a tab from my clipboard into a line of code, but not at the end, e.g. &quot;int i = 0; // init i&quot;. If I want more space between the command and the comment and I use my tab key, then tabs get inserted properly, but if I put a tab into my clipboard and insert it with Ctrl-V, then ALL tabs get replaced with spaces.

This problem does not occur in C++ files of the same solution, even though I set the tab configuration identical for all languages.

Cheers
Peter

_This issue has been moved from https://developercommunity.visualstudio.com/content/problem/154651/vs2017express-c-tabs-replaced-with-spaces.html
VSTS ticketId: 528589_
_These are the original issue comments:_

Peter Meier on ‎11‎/‎23‎/‎2017, 02:27 AM (41 days ago): <p>Also I just found out that if I hit the auto format keys (Ctrl-K, Ctrl-D), all the tabs I inserted between the command and the comment are replaced with spaces again. I searched through Tools/Options/Text Editor/C#/Code Style/Formatting, but I found no way to change this annoying behavior. I don't understand why C# won't let me align my comments the way I want them to be when it's no problem for C++.</p>

Jinu Joseph [MSFT] on ‎12‎/‎20‎/‎2017, 02:48 AM (14 days ago): <p>We appreciate you taking the time to report this problem. We are currently prioritizing problems that are impacting a broad set of our customers, so we may not be able to investigate this one immediately. We know this problem is important to you, so we will continue to monitor it.</p>

_These are the original issue solutions:_
(no solutions)

**Comment by jcoutch** [comment]
I'm having a similar problem with Visual Studio 2017 Professional (15.7.0.)  I have my indentation style set using `.editorconfig`:

```
[*.cs]
indent_style = tab
indent_size = tab
tab_width = 4
```

If I add a newline to a C# file, tabs are inserted as expected.  But, if I format the document using Ctrl+K Ctrl+D, any tabs in the file are replaced with spaces.

If I use ReSharper to format my document, all indentation is converted to tabs as expected.

**Comment by FlorianHaupt** [comment]
Would be nice to get this fixed, would save much time.

**Comment by elliott-beach** [comment]
I was thinking of submitting a repro, but the OP hits the nail on the head:
>  just found out that if I hit the auto format keys (Ctrl-K, Ctrl-D), all the tabs I inserted between the command and the comment are replaced with spaces again. I searched through Tools/Options/Text Editor/C#/Code Style/Formatting, but I found no way to change this annoying behavior. 

This creates big whitespace changes after formatting the document as we use tabs at work. Please fix this.

**Comment by sharwell** [comment]
📝 The issue here is hard tabs contained *within* a line of code, i.e. tabs not used for indentation.

**Comment by kendrahavens** [comment]
Design meeting notes:
In the past we've had conflicting requests on the behavior for tabs within a line of code so we need to add an option for this behavior.

My proposal:
* add editorconfig setting that enforces tabs within a line of code. `tab_within_line`
* add new toggle in **Tools > Options > C# > Tabs**: Enforce tabs within lines.
  * It could also go under **Tools > Options > C# > Code Style > Formatting > Spacing** if that is the preferred place to keep editorconfig settings.

@heejaechang @sharwell

**Comment by FlorianHaupt** [comment]
@kendrahavens 
Nice proposal to add an extra option.
> Design meeting notes:
In the past we've had conflicting requests on the behavior for tabs within a line of code so we need to add an option for this behavior.

Are there really users that like to keep tabs only on specific parts, like indent?
![tabs-settings](https://user-images.githubusercontent.com/18165738/49427897-d7adf800-f7a4-11e8-9842-6077bbb51952.png)
As the current option does reflect what it should do, to "keep tabs" and do not replace them with spaces - This is the behavior as I know it from previous versions and as it is within the C++ text editor - I would prefer an new option that says "keep tabs only for indent" for this new behavior. One saying "Enforce tabs within lines" would still confuse and let me wonder why I have to check that an extra option in situations I already selected "Keep tabs".

**Comment by sharwell** [comment]
The preliminary design discussion is now complete. We will review the final user experience once it is ready.

**Comment by levicki** [comment]
> The issue here is hard tabs contained within a line of code, i.e. tabs not used for indentation.

I keep hearing people referring to tabs as hard when it's actually the other way around -- tabs are soft (as in you can replace them with any number of spaces on display),, and spaces are hard (as in hard-coded number in the file itself which you can't reformat so easily on display).

That said, I would really prefer if Tab key did what it says on the tin -- inserted a Tab character into the editor. Currently in Visual Studio 16.5.4 that doesn't seem to be the case even though I have configured everything to have Tabs instead of spaces.

I'd like to voice my displeasure towards two trends going on in Visual Studio editor since VS 2015:

1. Editor trying to be "smart" and disregarding user input more and more (i.e. I enter Tab and it enters spaces even though it's clearly configured to use Tabs)
2. Premature code analysis and error checking (i.e. telling you your code is broken and offering to "fix" it before you ever get a chance to complete it)

Those additions waste enormous amounts of developer time and effort on fighting them when they are wrong (and sadly they can be wrong a lot). The least you could do is offer an option to disable both behaviors globally.

Finally, I'd appreciate if there was some workaround for this spaces instead of tabs issue, it's driving me nuts.

**Comment by Edgs** [comment]
This is still a problem.  The older versions of editors never used to behave like this.  Other editors don't behave like this.  Even Notepad doesn't behave like this.

I'll explain simply:
With the option to KEEP TABS selected in the Editor settings, TABs typed within code (so end of line comments line up, for example.) are later changed into spaces by Visual Studio.  When copying & pasting a line, for example.
I NEVER want this to happen.  That's why I select the option to KEEP TABS.
If I type a TAB in my code, I NEVER want it changed it to spaces.  It's really that simple.

Why is this not fixed after 2+ years?
This is not an enhancement request, but simply a request that the editor behave as editors have always behaved since the dawn of computing.  At some point, someone in Microsoft decided to 'muck' around with users typed code and change it, when not one user asked for that to happen.

> > Are there really users that like to keep tabs only on specific parts, like indent?
> Yes, it's actually a primary request for users who indent with tabs.
That's simply not true.  Show us ONE request from ANYONE who asked for their TABs to be changed to spaces within their code when using the option KEEP TABS.

Also worth noting that this still happens when the 'Use Adaptive Formatting' option is turned off.  So it is impossible to prevent this from happening.

**Comment by sharwell** [comment]
> Why is this not fixed after 2+ years?

There is no one consensus on the definition of correct behavior. The overwhelming majority of users are happy with the current implementation of tabs/spaces handling. Accounting for the remaining ones without breaking the experience for users who are happy with the current behavior is a particularly challenging exercise that requires both design and implementation work.

I'm not sure this will move up on the internal priority list in the near future, but if an external user wanted to spend the time to define and implement the full experience we would be happy to review it. See #23394 for a great example of a feature which shipped because a contributor went through this process. 😄

**Comment by levicki** [comment]
@sharwell 
> There is no one consensus on the definition of correct behavior.

How about you take the same approach as Git does for line endings?

1. Replace TABs with spaces on reformat
2. Replace spaces with TABs on reformat
3. Leave both as entered on reformat

Shouldn't the above satisfy everyone involved?

Come on people, this is not rocket science, it's a text editor for $(Deity)'s sake!

**Comment by CyrusNajmabadi** [comment]
I've never seen git do that.

**Comment by Edgs** [comment]
> 
> 
> I've never seen git do that.

No matter, you've never seen any other existing or previous code editor, since text editors first existed, behave the way VS does.
When auto-indenting, it ALWAYS changes TABs between code and line comments to SPACES no matter what settings are selected - like the 'always keep TABs' setting.
Nobody asked for this behaviour.  It's a bug.
Obviously it's more convient to blindly deny it here rather than simply fix it.

**Comment by levicki** [comment]
> I've never seen git do that.

How about not commenting before understanding what you read? I never said Git does anything with spaces and TABs -- I explicitly mentioned line endings.

**Comment by sharwell** [comment]
> Nobody asked for this behaviour. It's a bug.

This is not correct. Some users who use tabs for indenting prefer the current setting. _Most_ users who use tabs for indenting either prefer the current behavior _or_ they are OK with either behavior.

We are willing to reconsider the design since some teams still aren't happy with the current behavior, but only if the design provides a comprehensive approach that works for both preferences.

> How about not commenting before understanding what you read? I never said Git does anything with spaces and TABs -- I explicitly mentioned line endings.

Git has the ability to transparently normalize line endings as part of commit/checkout, but there are many reasons why this would not work with tabs/spaces:

1. Git doesn't have the ability to alter its understanding of normalization to include characters other than end-of-line characters.
2. Not all users are working with Git for source control.
3. Tools which operate on files with checksum validation (e.g. debugging) _explicitly_ account for the fact that every text file has two possible forms: one with `\r\n` and a second with `\n`. All of these tools will break if new normalization characters are added.

**Comment by CyrusNajmabadi** [comment]
> How about not commenting before understanding what you read?

How can I understand without discussing and talking with people about the topic?

**Comment by CyrusNajmabadi** [comment]
So, i would need some strong answers to problems i see arising from changing the representation of the file for different developers.  First, that would violate some of our efforts we have around reproducible builds.  Second, it seems like it would just cause problems for normal situations like:

```c#
var v1 = from c in customers
         where c.Age >= 21
         ...
```

Here, the continuation lines need to be indented 9 columns to maintain alignment.  Replacing these with tabs just breaks this.  Even if it was on some tab multiple for some developers, it might not be for others.

In general, I think all developers on the team (including CI) should operate on a bit-for-bit identical version of the code for many important reasons.

**Comment by levicki** [comment]
@sharwell 
> Git has the ability to transparently normalize line endings as part of commit/checkout, but there are many reasons why this would not work with tabs/spaces...

Let me be absolutely clear that I did not suggest changing Git behavior, but rather emulating their set of options for the particular problem at hand.

With that out of the way, can we agree that VS editor already does some sort of "normalization" of TABs and spaces?

The problem in my opinion is twofold:

1. The existing settings that govern TAB .vs. space editor behavior are all over the place and totally do not work as described in the UI, much less produce results that are expected.
2. There is no `What You Type Is What You Get` editor setting -- I just want (and I am sure I am not the only one) that when I enter mixed TABs and spaces on a single line of code that they stay exactly like I entered them (save for expression reformatting which is already customizable enough for everyone's taste).

@CyrusNajmabadi 

> How can I understand without discussing and talking with people about the topic?

You butted into the middle of a discussion by putting words in my mouth because you haven't bothered to understand what I wrote. Maybe that is somehow my fault because English is not my primary language and what I wrote is hard to understand, but in all honesty I am not sure what are you arguing for (or against). If you want to keep spaces in your files then rest assured that nobody here wants to take that away from you. I am asking for more options, not less.

So once again, I only want an editor option to not touch beginning of line indentation, variable name indentation, and comment indentation -- it is irrelevant whether those are done with TABs, spaces, or a mixture of the two -- I want it preserved.

I am really surprised to hear that something like that cannot be done while still keeping all current options.

**Comment by CyrusNajmabadi** [comment]
> You butted into the middle of a discussion by putting words in my mouth

I didn't.  I said i'd never see such behavior.  I have not.  That was me just giving insight that I wasn't familiar with what you are talking about.

**Comment by CyrusNajmabadi** [comment]
> I am not sure what are you arguing for (or against)

I'm happy to clarify any of my points.  Def ask questions and I'll get back to you asap.  If you'd like another venue to discuss things (perhaps in realtime) I'm also happy to use gitter or discord to continue the discussion :)

**Comment by CyrusNajmabadi** [comment]
> I am asking for more options, not less. 

Sure.  However, I'm pointing out that those options are potentially quite problematic.  I would be wary about adding them without fully understanding all the implications here.  And to get there, i need to talk about the topic.

> Therefore, please stop diluting the topic with pointless and/or obvious comments.

It was not obvious to me.  That's the reason i made the comment.  I was unfamiliar with what you were talking about, so i pointed that out to get clarity.

**Comment by CyrusNajmabadi** [comment]
> I am really surprised to hear that something like that cannot be done while still keeping all current options.

The challenge is that you may be asking for X, and we may provide *exactly* X, only to find out soon after that you actually wanted X', and that X is not sufficient for all your cases.  THe purpose of the discussion, and the examination of all the tangential concerns is to try to get as full an understanding as possible so as to provide the best possible solution here.  

As you may have noticed, this is actually a fairly complex space with lots of varying concerns.  You may have even noticed that different people have asked for things *in conflict* with other requests.  It's not as simple as just taking one proposal and implementing it and leaving it at taht.

**Comment by levicki** [comment]
> The challenge is that you may be asking for X, and we may provide exactly X, only to find out soon after that you actually wanted X', and that X is not sufficient for all your cases

So now you are implying that I don't even know what I want? What's next, telling me to get used to doing things your way because reasons?

Well I know one thing for certain -- I do not want an AI or any piece of software to "think" instead of me and attempt to interpret my consistent keyboard input in different ways at different times as it pleases. Humans are creatures of habit and muscle memory is how we free our brain's capacity so it can perform more complex tasks, contextual controls or behavior are breaking muscle memory and are thus bad UI and software design.

Again, all I want is that spaces and tabs are left as I entered them when the line is reformatted. Nothing more, nothing less. All existing indentation and spacing formatting was fine for me before this "smart" tab to space conversion on reformat was added. I just want an option to turn it off.

How hard can that be?

**Comment by CyrusNajmabadi** [comment]
> So now you are implying that I don't even know what I want?

Yes.  That's always a possibility.  It's happened numerous times in the past with many requests from many customers.  It's always something that needs to be kept in mind.  Furthermore, it may be what you want, but it may be wrong for other users complaining about something similar.  So navigating all the potential requests and constraints here can be tricky and takes a lot of communication and a lot of thought.  Thanks! :)

**Comment by Edgs** [comment]
> As you may have noticed, this is actually a fairly complex space with lots of varying concerns...

Then just concentrate on the **original problem in the first post**.  It's a very straightforward problem.
**Tab characters, between code and double-slashed comments, are changed to space characters when auto-indenting is performed.**
Auto indenting works fine - it uses TAB characters when asked, but TAB characters after the code are changed.
Why?

This is with **Always keep TABs** option selected.
If this is a feature, and not a bug, then someone please explain why, when **Always keep TABs** is selected, these TABs are being replaced by Spaces?

This behaviour does not exist in older versions of VS.  At some point it has been added by someone, for some reason.
That was the point at which the user experience was broken.

**Comment by CyrusNajmabadi** [comment]
> How hard can that be?

It could potentially be very difficult.  The formatting engine is one part of Roslyn that i find the most challenging.  Not "it's one of the top challenging parts".  Rather: "it is literally the most challenging pieces of roslyn".

Changes here often have very unexpected effects that are unintended and can break users who have become accustomed to how things work.  We have to be very delicate here as this can be very detrimental to codebases.

**Comment by CyrusNajmabadi** [comment]
> Auto indenting works fine - it uses TAB characters when asked, but TAB characters after the code are changed.

I don't think this is the case.  For example, i believe we may choose to align things, and not stick with tabs.  But i would have to go check on that.

**Comment by CyrusNajmabadi** [comment]
> This is with Always keep TABs option selected. If this is a feature, and not a bug, then someone please explain why, when Always keep TABs is selected, these TABs are being replaced by Spaces?

I believe this may be a bug or a feature depending on ones perspective.  I certain see how it can be a bug.  However, that doesn't mean that just because it's a bug that it is simple to fix or that it is without consequences if changes are made here.  Again, as i mentioned above, the existing formatting system is highly complex (including with likely bugs in it that many have taken dependencies on).  So changes need to be very judiciously made.

**Comment by CyrusNajmabadi** [comment]
> This behaviour does not exist in older versions of VS. At some point it has been added by someone, for some reason.
That was the point at which the user experience was broken.

That's quite possible.  My guess is that it happened when we did the entire roslyn rewrite.  We tried to preserve a lot of old behavior, but we very likely did not given the huge complexity in the old system and that the new system takes an entirely different approach on things.  Unfortunately though, we've now had this new system a long time, and we are wary about subjecting people to more potential changes in behavior, esp. as it might fix this issue, only to cause problems for others :-/   It's a tricky situation to be sure!

**Comment by CyrusNajmabadi** [comment]
Here's an example of an issue, based on previous discussions with other customers on this topic.  What i've heard from *other* customers (again, not you), is that it's not necessarily that they want `tab` to be preserved, but that they want comment alignment to be maintained.  in a system where we kept the tabs, we could potentially end up with something like this happening:

```c#
    x();    // this is here
    y + z + w + x();    // but now this is here
    x();    // but this is back down here.
```

And so on and so forth, instead of keeping comments aligned as they wanted.

**Comment by Edgs** [comment]
> I believe this may be a bug or a feature depending on ones perspective. I certain see how it can be a bug. However, that doesn't mean that just because it's a bug that it is simple to fix or that it is without consequences if changes are made here. Again, as i mentioned above, the existing formatting system is highly complex (including with likely bugs in it that many have taken dependencies on). So changes need to be very judiciously made.

I don't disagree with that, but there seems to be great resistance to even acknowlege that the behaviour has changed at some point, and that it doesn't seem correct, from a users point of view, that selecting **Always keep TABs** means **Always keep TABs except sometimes they will be changed to spaces**

**Comment by Edgs** [comment]
> 
> 
> Here's an example of an issue, based on previous discussions with other customers on this topic. What i've heard from _other_ customers (again, not you), is that it's not necessarily that they want `tab` to be preserved, but that they want comment alignment to be maintained. in a system where we kept the tabs, we could potentially end up with something like this happening:
> 
> ```cs
>     x();    // this is here
>     y + z + w + x();    // but now this is here
>     x();    // but this is back down here.
> ```
> 
> And so on and so forth, instead of keeping comments aligned as they wanted.

But this situation is _worse_ when spaces have been inserted.  This is my biggest frustration with this change in behaviour, and there is no way to avoid it.  It's far easier, and far less typing, to manually realign with tabs than with spaces.  To make matters worse, some lines are left alone & some are changed to spaces, so you don't know until you start typing.  It's VERY unpredictable and frustrating.

**Comment by CyrusNajmabadi** [comment]
> But this situation is worse when spaces have been inserted. 

I'm not sure i agree.  Spacing allows thigns like alignment to happen in a very flexible fashion that is independent of whatever tab settings user have.  It may be worse for you, but better for others.  Hence the dilemma :)

**Comment by CyrusNajmabadi** [comment]
> It's far easier, and far less typing, to manually realign with tabs than with spaces.

Oh, i've never found that to be the case myself.  This is why i always use spaces.  I can perfectly realign on any column effectively with spaces.  Ymmv of course :D    

> It's VERY unpredictable and frustrating.

Agreed.  This is one reason i ban tabs in all codebases i control.  It's just far too unpredictable across all scenarios (esp. when working with multiple tools).  I get that may not be palatable for your own preferences though of course :-/

**Comment by Edgs** [comment]
You'd see things very differently if on some lines of your code, your spaces were constantly being replaced by tabs.

**Comment by levicki** [comment]
@CyrusNajmabadi

99% of what you wrote translates into excuses to not even attempt to change anything because something could break while at the same time you are totally disregarding that it has already been broken for us, hence the bug reports.

The remaining 1% is barely contained smugness and superiority of someone in a position of some power, and it is clear that you enjoy imposing your way on others given your "ban on tabs" which you mentioned above.

So far you have not addressed any of my points, instead you keep spouting some weird mix of anti-tab religious zealotry and corporate PR talk.

At the risk of getting banned from commenting on this and every other Microsoft repo I am going to call you out on your bullshit -- you, Sir, are full of it and presenting rational arguments to you is a colossal waste of time.

i am out of here, and I will be advocating in my company to drop Visual Studio licensing. Maybe you will learn to honor user's preferences when they hit you in the wallet.

**Comment by sharwell** [comment]
We'll give this a bit of time to cool down. I did try implementing https://github.com/dotnet/roslyn/issues/24031#issuecomment-444067640 once in the past, but it was difficult to get the formatter to understand the leading whitespace on a line might contain both hard tabs (indent) and spaces (alignment).

Even the simpler option of having a mode to toggle between the current behavior and the pre-Roslyn behavior would be very difficult to implement. Since the formatter applies tab/space conversion in the middle of lines, it's unfortunately not possible to take the "easy" approach of limiting a fix to the handling of indentation.

The current workaround is to disable the auto-format on type/paste option:

![image](https://user-images.githubusercontent.com/1408396/108928995-0fa8ec00-75f8-11eb-807e-8b760c27698f.png)

Altering the behavior within the formatter is very challenging (combinatorial explosion in the heuristics frequently lead to subtle regressions) so given the behavior is 6 years old with less than 10 👍 votes, it's going to be difficult to prioritize a change.

**Comment by CyrusNajmabadi** [comment]
> So far you have not addressed any of my points, instead you keep spouting some weird mix of anti-tab religious zealotry and corporate PR talk.

If i've missed any points, i apologize.  There's a lot being discussed here, and i likely missed it.  Feel free to point it out once this is unlocked and i'll respond as best as i can.

> because something could break while at the same time you are totally disregarding that it has already been broken for us, hence the bug reports.

i'm not disregarding this.  I'm just providing some insight that sometimes it's better to just stick with a regression versus introducing multiple regressions at multiple points in time.  i definitely commiserate with the initial regression being unpleasant.  However, with a rewrite the size of roslyn *and* the entire formatting engine being rewritten *from the ground up* it was invariable that this would happen to some of these cases.  However, now that we have the current system, i would be loathe to potentially have this same thing happen *again*.

I totally get that that can be frustrating.  A group of people (that you are not a part of) is having their needs met, while yours are not.  However, this really is a space where potential changes risk very bad outcomes for many groups.  As i mentioned in https://github.com/dotnet/roslyn/issues/24031#issuecomment-784574391, this code is *the* most complex in the entire codebase for me.  Changes to it are both non-trivial and carry a large amount of risk.  Just a few days ago i discovered something new and unexpected about it for me, and i had to completely give up on a refactoring here because there was a part of this whose behavior was too subtle and complex to fully understand and i had to timebox myself out of it.

--

Again, I'm happy to help clarify any of my statements, or address any points i may have missed.  My hope was to try to provide some clarity on the challenges here as this isn't really a case of: "oh, this is just a bug with a trivial solution" but rather more like "this is a potential mine-field that we have to be very careful of".  

Also, as per https://github.com/dotnet/roslyn/issues/24031#issuecomment-784436432, i would be happy to continue this discussion in realtime on discord or gitter.  It might be more conducive than here as we can cover a lot more stuff, and we can make sure any particular questions/concerns you have are responded to.  If you are interested, LMK and i'm happy to meet up with you at practically any time.  thanks!

**Comment by jmarolf** [comment]
I can certainly understand the frustration of **Always keep TABs** not doing what it says on the tin! I truly do want to fix this as we have under-served tabs users in the past and want to ensure it is just a viable and useful as spaces. 

Let's talk potential solutions (setting aside for a moment the engineering cost) and see if we can agree on which one would be most helpful. Once we agree on a design, we can evaluate the engineering difficulty of it.

## Tabs only mode

Simplest solution would be to have an option that means "if this option is _on_ Tabs are never allowed to be converted to spaces by the formatter". I will tentatively call this _Tabs only mode_ (since sadly "Always" is taken). I think this will operate like folks want but there is a corner case that has been called out that we need to consider which is what if we can't perfectly align things with just tabs?. So, if you set your tabs to some indenting (like say "4") we would try to get you as close as we can based on those settings with the understanding that we could be off by one or more. This would be an example of a worst-case scenario where we could choose to either be off by _one_ or _three_ when tabs have an indent size of 4.
```csharp
void Method()
{
→   var·data·=·from·value·in·new[]·{·1,·2·}
→   →   →   select value; // off by three
}
```

```csharp
void Method()
{
→   var·data·=·from·value·in·new[]·{·1,·2·}
→   →   →   →   select value; // off by one
}
```

I think intuitively we would say "We should be off by as little as possible". So, for worst case scenarios like the ones above we would choose option 2 over option 1. The question for this mode is can the developer ever actually get into a good state when things are not misaligned in this worst-case? The answer is probably not in this mode. Say I add an additional space in front of `from`:
```csharp
void Method()
{
→   var·data·=··from·value·in·new[]·{·1,·2·}
→   →   →   →   select value;
}
```
The formatting engine will just take it away again as we already have a preference that keeps the spacing here to just one space. You could use spaces at the beginning of the line to align things and they would be left alone:
```csharp
void Method()
{
→   ·var·data·=·from·value·in·new[]·{·1,·2·}
→   →   →   →   select value;
}
```
 But that is dangerously close to _mixing_ tabs and spaces which is the very thing we don't want to do. I think like it's not obvious to me what would be most excellent here. But I am just one guy what do I know? Maybe we try a vote?
 
- **React with 🚀 for _Tabs only mode, stop indenting, never overshoot_**
```csharp
void Method()
{
→   var·data·=·from·value·in·new[]·{·1,·2·}
→   →   →   select value; // off by three
}
```
- **React with 🎉 for _Tabs only mode, choose the smallest number to be off by on alignment_**
```csharp
void Method()
{
→   var·data·=·from·value·in·new[]·{·1,·2·}
→   →   →   →   select value; // off by one
}
```

## Tabs preferred mode

An alternative would be "Use spaces as little as possible" named something like _Tabs preferred mode_. In this mode we would use tabs to re-align things and only use spaces iff there was no other means to get things to line up.

```csharp
void Method()
{
→   var·data·=·from·value·in·new[]·{·1,·2·}
→   →   →   ···select value;
}
```

- **React with :+1: for _Tabs preferred mode_**

**Comment by levicki** [comment]
@sharwell 

I apologize for being rude, but ever since Visual Studio 2015 I am sick of Visual Studio messing with my keyboard input and not respecting my configuration choices (and in some situations not even giving me any choice).

I understand that the problem with mixed TABs and spaces such as this very contrived example:
```
	if ((value ==  8) &&					// NOTE: Don't mess with comment alignment too
	    (  age == 21) &&					//       or someone will be really annoyed
	    ((flags & DONT_MESS_WITH_TABS) ||			
	     (flags & DONT_MESS_WITH_SPACES_EITHER))
	   ) {
		// TOOD: figure out how to handle this...
		abc.Value		= 3;
		abc.AnotherValue	= "MaybeNotTouchThis";
	}
```
Are really complex to solve, but perhaps it would be prudent not to mess with indentation when mix of tabs and spaces is detected at line start? The rest of the line could be formatted according to the rules, leaving anything past the expression end as-is to avoid messing up comment alignment.

I understand that perfection will never be possible, but at least having an option to disable auto-conversion of tabs to spaces would be a start.

**Comment by jmarolf** [comment]
@levicki thanks for this.
```csharp
void·function()
{
→   LongClassName→   c;→   →   →   →  	//·tabs·here
→   HRESULT→   hr = E_FAIL;

→   if·((value·==··8)·&&→   →   →   →   →   //·NOTE:·Don't·mess·with·comment·alignment·either
→   ····(··age·==·21)·&&→   →   →   →   →   //·······or·someone·will·be·really·annoyed
→   ····((flags·&·DONT_MESS_WITH_TABS)·||			
→   ·····(flags·&·DONT_MESS_WITH_SPACES_EITHER))
→   ····)·{
→   →   // TOOD: figure out how to handle this...
→   →   abc.Value→   →   = 3;→   →   →   //·and here
→   →   abc.AnotherValue→   =·"MaybeNotTouchThis";
→   }
→   	
→   EvenLongerClassName·e·=·new·EvenLongerClassName();→   //·but·not·here
}
```

In the sections where we are using tabs to align comments. If we pressed enter a the ⌶ location what would you expect the formatter to do ideally?
```csharp
→   ····(··age·==·21)·&&→   →   →⌶   →   →   //·······or·someone·will·be·really·annoyed
```

**Comment by Edgs** [comment]
> In the sections where we are using tabs to align comments. If we pressed enter a the ⌶ location what would you expect the formatter to do ideally?

THE problem currently is that if you press enter on that line, or a line next to it, and a code or comment alignment takes place, all those tabs before the comment are replaced by spaces.  Every time, no matter what TAB options are selected.
I don't think there's anything wrong with the 'alignment help' as it is, except that it ONLY uses spaces, and never tabs, as reqested by the user.

Please fix this.

**Comment by levicki** [comment]
> In the sections where we are using tabs to align comments. If we pressed enter a the ⌶ location what would you expect the formatter to do ideally?

Is there any particular reason why I would want to press Enter in the middle of those tabs?

If I wanted to move that comment to the next line I would probably press Enter right after last non-whitespace character, not in the middle of the whitespace. 

That aside, I would expect it to indent the comment with tabs from the start of line, and if it cannot decide whether to also add spaces to align the comment to the same column as the previous line then just leave the caret before the first non-whitespace character so I can do manually whatever I want.

If you intentionally wanted to split the line so that there are two tabs stuck to the coment going to the next line then I really don't know the answer because I'd never do that unless there was no auto-indent and unkess previous line was indented with exactly two tabs. Even then I'd still have to clean up the trailing whitespace on the previous line if I did that so that would make it even less likely.

As it was stated before, we don't need perfect solution.

Hovewer, I am afraid that we have gone way too far in our attempts to enable software to predict and adapt to human behavior  -- to the point of making the software behavior itself unpredictable as a result.

The worst part for me is that we can't turn it off. All I want from software is:

1. Always respect user configuration
2. Never disregard user input
3. Behave consistently and reliably

The third point is really important because it enables muscle memory which frees brain processing capacity for the task at hand (i.e. writing quality code instead of fighting with the editor).

To put it short - follow the principle of the least astonishment.

**Comment by sharwell** [comment]
> Is there any particular reason why I would want to press Enter in the middle of those tabs?
>
> If I wanted to move that comment to the next line I would probably press Enter right after last non-whitespace character, not in the middle of the whitespace.

In most cases, pressing <kbd>Enter</kbd> in the middle of whitespace already behaves the same as pressing enter at the end of that same whitespace. This feature is frequently used.

> That aside, I would expect it to indent the comment with tabs from the start of line, and if it cannot decide whether to also add spaces to align the comment to the same column as the previous line then just leave the caret before the first non-whitespace character so I can do manually whatever I want.

Can you show this using the `→` and `·` characters?

**Comment by levicki** [comment]
@sharwell 
> In most cases, pressing Enter in the middle of whitespace already behaves the same as pressing enter at the end of that same whitespace. This feature is frequently used.

Funny, I never press Enter at the end of whtespace, but at the beginning.

> Can you show this using the → and · characters?

I can try, in both examples below I assume pressing Enter at the beginning of whitespace.

**Version 1 (VS editor cannot figure out whether I also want spaces or not after initial tab alignment)**
```
→   if·((value·==··8)·&&
→   //·NOTE:·Don't·mess·with·comment·alignment·either
   ^- cursor here after pressing Enter so I can align further manually if I want
→   ····(··age·==·21)·&&→   →   →   →   →   //·······or·someone·will·be·really·annoyed
```
**Version 2 (VS editor has gained awareness and is becoming Skynet by rewriting its own code as we speak)**
```
→   if·((value·==··8)·&&
→   ....//·NOTE:·Don't·mess·with·comment·alignment·either
→   ····(··age·==·21)·&&→   →   →   →   →   //·······or·someone·will·be·really·annoyed
```
But that still focuses on minor style details while not addressing the key complaint most of us tab users here have, namely:

**When I type this:**
![image](https://user-images.githubusercontent.com/16415478/111296459-ce32bc00-864c-11eb-80f3-825f02123b15.png)

**I don't want it being changed to this when I type `;` at the end of line:**
![image](https://user-images.githubusercontent.com/16415478/111296543-e7d40380-864c-11eb-82c3-64b9475770c3.png)

Yes I know I can disable auto-format on `;`', but I want other formatting to happen, just not that one. If you just fixed that, you would have much less complaints to begin with.

**Comment by Edgs** [comment]
Personally, I would be happy for no comment alignment to take place at all.
Code statement alignment is generally predicatable, and is mostly determined in the user options.  The Editor can make reasonable assumptions about indentation if it is formatting code.
Comment alignment is completely arbitrary - for example I sometimes like to indent my comments, like pseudo-code, or sometimes have them all vertically aligned.  The only thing the Editor can do with this is move it somewhere I don't want, or just leave it alone.

For example, currently I can copy and paste a block of code, and the Editor will keep the same predictable alignment for the code, but the comments will be moved in an attempt to align them with nearby comments.  I don't want this to happen.  The Editor is just guessing in this case, and the behaviour is unpredictable.
It's all made worse by the Editor only using spaces, so I have to go back & find & remove those to realign the comments as I see fit.  In 9 times out of 10 it just creates more typing.

**Comment by levicki** [comment]
@jmarolf @sharwell 

Here is another example of the undesired behavior I detailed with screenshots in my previious comment above:

![image](https://user-images.githubusercontent.com/16415478/111482241-019c4600-8734-11eb-8615-5020a6c5723b.png)

Note that in my root Visual Studio Projects folder I have an `.editorconfig` file which contains, among other things:
```
root = true
...
dotnet_diagnostic.IDE0055.severity = none
```
However, it seems that IDE0055 configuration is being entirely ignored, and clicking on any of the **Suppress or Configure issues** sub-options for it has literally no observable effect (no GUI feedback, no file changes / creations, nothing).

**Comment by sharwell** [comment]
> However, it seems that IDE0055 configuration is being entirely ignored

Something is overriding this value (most likely a .editorconfig file between that location and the location of the file also contains `root = true`, which prevents further inheritance).

**Comment by levicki** [comment]
> Something is overriding this value (most likely a .editorconfig file between that location and the location of the file also contains `root = true`, which prevents further inheritance).

There is no such file between.

**Comment by sharwell** [comment]
Can you file a separate issue for that? Disabling a diagnostic is expected to disable a diagnostic. 😄

**Comment by levicki** [comment]
> Can you file a separate issue for that? Disabling a diagnostic is expected to disable a diagnostic. 😄

There is already an issue for IDE0055 about it fighting vertical alignment (i.e. replacing tabs with spaces as I showed above). It's open for God knows for how long just like this one. I am not even going to mention the one I reported in .Net Environmet namespace two years ago.

Do you honestly believe that opening another issue will help? At this point I am discouraged to report more because nothing ever seems to get fixed (usually because "it's complicated").

**Comment by sharwell** [comment]
> There is already an issue for IDE0055 about it fighting vertical alignment ... Do you honestly believe that opening another issue will help?

Yes, you're describing a bug related to a feature much bigger than just IDE0055 that we rely on ourselves.

https://github.com/dotnet/roslyn/blob/c18da5a9e58e02ba186197c221815ead102ff41e/.editorconfig#L144-L145

**Comment by CyrusNajmabadi** [comment]
> At this point I am discouraged to report more because nothing ever seems to get fixed (usually because "it's complicated").

We do fix hundreds of issues.  But there are a lot of things out there that people want fixed.  So it's definitely possible that the set you care about isn't the set that always makes it into a release.   The ones in this area are particularly thorny, and we lack a good mechanism to get high confidence that changes here not only make things better for these cases, but also do not make things worse for cases that others care about.

**Comment by FlorianHaupt** [comment]
This is really worse, what I read here. This misbehavior on tabs was reported shortly after the release of the VS2015 version, multiple times, by different people. Most of this reports were closed after a short period. I am quite sure, that everyone here knows about the license costs - I use a lot of cheaper software where the companies do not deal in such a bad way with their customers and fix such bugs when they get reported. Funny enough that this is open, the one reported on VS2017 Express.
You developed an new auto format engine and it fails to give the same good experience as the one that was used before and instead of fixing it, you prefer to spend hours of discussion with arguments that have nothing to do with the given issue. This is not understandable, a new engine has bugs and you do not feel the need to get it fixed? On top it is for the comfort in code editing and this is a bug that breaks the comfort in editing, creates unnecessary extra work – a very worse experience.
How much patience do you expect, by people that have to deal with this bug on every day?
I still remember that there once was really good bug handling within the visual studio team - seems that this is history.

**Comment by sharwell** [comment]
@FlorianHaupt For 75% of my career to date, I was a non-Microsoft developer working on projects primarily using tabs for indentation. The current behavior better represented the expected behavior for those projects, so on those projects I would have considered VS2015 to be the point where the bug was _fixed_, not introduced. The complexity here is we're not talking about "fixing a bug", we're talking about introducing an option at a location that was never designed to be an option, and somehow getting both¹ behaviors to work seamlessly.

¹ This is a simplification, because in reality there are _at least_ three very different behaviors being requested.

**Comment by Edgs** [comment]
> 
> 
> @FlorianHaupt For 75% of my career to date, I was a non-Microsoft developer working on projects primarily using tabs for indentation. The current behavior better represented the expected behavior for those projects, so on those projects I would have considered VS2015 to be the point where **the bug was _fixed_**, not introduced (note that I was not working at Microsoft for the VS2015 release where this behavior changed, so I was just a user). The complexity here is we're not talking about "fixing a bug", we're talking about introducing an option at a location that was never designed to be an option, and somehow getting both¹ behaviors to work seamlessly.
> 
> ¹ This is a simplification, because in reality there are _at least_ three very different behaviors being requested.

What 'bug' are you talking about?
Please read the initial post again, and please stick to one topic per thread.
There is nothing complex about the BUG he is reporting.  If that _is_ what you are talking about, then please demonstrate another editor in existence which replaces tabs with spaces between code and comments.  That is NOT expected behaviour.

If you are talking about something different then please split this thread as several quite different topics appear to be being discussed at once.  Nothing will get fixed that way.

**Comment by sharwell** [comment]
@Edgs I'm referring to the same issue reported above. I'm not aware of another editor that works like this. When Visual Studio 2015 was released, I was happy to see it was the first editor to finally fix the behavior here.

**Comment by levicki** [comment]
> We do fix hundreds of issues. But there are a lot of things out there that people want fixed.

That is just a proof that Agile (a.k.a. move fast and break things) approach to development doesn't really work as well as some people claim it does, and that internal QA team is necessary for development of quality software.

> So it's definitely possible that the set you care about isn't the set that always makes it into a release. The ones in this area are particularly thorny, and we lack a good mechanism to get high confidence that changes here not only make things better for these cases, but also do not make things worse for cases that others care about.

I apologize in advance for the offtopic, but I have to point you and @sharwell to an [issue](https://github.com/microsoft/dotnet-framework-early-access/issues/58) I submitted in .Net Framework regarding `Environment.SetEnvironmentVariable` function.

The original issue was initially moved to .Net Core and marked `won't fix` for .Net Framework, and it was archived in the meantime. I'd like you to spend some time if you can reading the back and forth arguments there.

After you do that, please proceed to the [issue](https://github.com/dotnet/runtime/issues/1442) they recreated for .Net Core initially, which has later been returned to .Net Framework runtime, and on March 28th it is going to be two years without resolution.

I'd also like you and @sharwell to spend some time to read my back and forth with Microsoft software engineers there.

You can see everything that needs improvement in the way issues are handled now at Microsoft -- from developer(s) not understanding the issue despite clear explanations, over them trying to find excuses not to change anything, over them not understanding the undelying platform behavior and original Win32 API, over developers of 3rd party libraries and tools chiming in saying issue should be fixed, to the issue resolution being endlessly postponed because "more discussion is needed" or "triage is needed" or whatever the excuse of the day is.

And today here we are approx. 6 years since the release of Visual Studio 2015 where the issue was introduced and you are telling me that you need more time to make sure someone is not negatively impacted while people using tabs are negatively impacted for 6 years and counting.

If you do bother to read those discussions as I asked you to, you will hopefully see how much of my personal time and effort I have spent on submitting the issue, and explaining things that Microsoft's own software engineers should have known better than me. In case what I am hinting at isn't clear, the difference is that I am a customer who is paying for the OS and dev tools licenses who is spending own free time to help improve things, and your colleagues are doing this on the salary and with this kind of response I am getting I have absolutely no incentive to contribute further. On the contrary -- I have the urge to drop Microsoft products and suggest to the company I work for that we do the same wherever we can.

TL;DR -- I did provide feedback, I answered all questions asked by @jmarolf and @sharwell, and as far as I am concerned I did more than enough to help you understand the problem we are having and how you might solve it. I hope you prove me wrong and actually fix this issue, but I won't be holding my breath.

**Comment by Edgs** [comment]
So no other editor exhibits this behaviour, no editor before 2015 exhibited this behaviour, people are here reporting it as a bug, but you're are suggesting that it fixed some problem endemic since the dawn of computing.  That's bemusing to me, but if you're happy with it, then lucky you.  For everyone else, it's tough, because they **can't turn it off if they choose to**.

I'm not suggesting the behaviour is changed with no way for any user to keep their preferred style (which is actually what happened in 2015), just that the user options behave as advertised.

The formatting applied from the beginning of the line to code statements _does_ follow the selected option - use spaces or tabs.
The formatting between code statements and comments does _not_ follow the selected option - it always uses spaces.
It appears that the user option is ignored in this specific location.  The location and 'tabulation' of the formatting is fine, just the characters used.

It is, lets say, disappointing that it sounds like Microsoft can't maintain it's code because it is too complex.  If you just told me that this shouldn't be addressed because I was a nobody and my opinion wasn't important, I'd be annoyed, certainly, but overall less concerned.

**Comment by sharwell** [comment]
I'm going to pause comments again since this is getting longer but not likely to reach a conclusion on the current path. In the spirit of open source transparency, @CyrusNajmabadi and myself are attempting to explain the history of the current issue as well as the many different types of input considerations we face when attempting to resolve the issue. We understand that the current state isn't optimal for everyone, but honest risk evaluation makes it clear that alterations to address this issue will consume substantial time and engineering resources (_always_ disproportionately large when working in the formatter) which directly translates to us not being able to use those resources to address other features, bugs, and overall product polish. As such, we have made the decision to leave this issue open while we work on other items that have broader customer impact.
